I hit another bug trying to test this:

```
Traceback (most recent call last):
  File "/Users/alancoding/.virtualenvs/ansible/lib/python2.7/site-packages/ansible-2.7.2.post0-py2.7.egg/EGG-INFO/scripts/ansible-inventory", line 118, in <module>
    exit_code = cli.run()
  File "/Users/alancoding/.virtualenvs/ansible/lib/python2.7/site-packages/ansible-2.7.2.post0-py2.7.egg/ansible/cli/inventory.py", line 181, in run
    results = self.dump(results)
  File "/Users/alancoding/.virtualenvs/ansible/lib/python2.7/site-packages/ansible-2.7.2.post0-py2.7.egg/ansible/cli/inventory.py", line 199, in dump
    results = json.dumps(stuff, cls=AnsibleJSONEncoder, sort_keys=True, indent=4)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 251, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 209, in encode
    chunks = list(chunks)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 434, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 442, in _iterencode
    o = _default(o)
  File "/Users/alancoding/.virtualenvs/ansible/lib/python2.7/site-packages/ansible-2.7.2.post0-py2.7.egg/ansible/parsing/ajson.py", line 66, in default
    value = super(AnsibleJSONEncoder, self).default(o)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: {u'openshift.io/scc': u'privileged'} is not JSON serializable
```

This bug appears to not exist in devel, but it does kind of prevent fully testing _this_ bug.

--- a/changelogs/fragments/49299-openshift-inventory-plugin-empty-pods-and-serialization.yaml
+++ b/changelogs/fragments/49299-openshift-inventory-plugin-empty-pods-and-serialization.yaml
@@ -1,0 +1,4 @@
+---
+bugfixes:
+- Convert hostvars data in OpenShift inventory plugin to be serializable by ansible-inventory
+- Fix error in OpenShift inventory plugin when a pod has errored and is empty

--- a/lib/ansible/module_utils/k8s/inventory.py
+++ b/lib/ansible/module_utils/k8s/inventory.py
@@ -114,17 +114,18 @@ class K8sInventoryHelper(K8sAnsibleMixin):
         for pod in obj.items:
             pod_name = pod.metadata.name
             pod_groups = []
-            pod_labels = {} if not pod.metadata.labels else pod.metadata.labels
-            pod_annotations = {} if not pod.metadata.annotations else pod.metadata.annotations
+            pod_annotations = {} if not pod.metadata.annotations else dict(pod.metadata.annotations)
 
             if pod.metadata.labels:
-                pod_labels = pod.metadata.labels
                 # create a group for each label_value
                 for key, value in pod.metadata.labels:
                     group_name = 'label_{0}_{1}'.format(key, value)
                     if group_name not in pod_groups:
                         pod_groups.append(group_name)
                     self.inventory.add_group(group_name)
+                pod_labels = dict(pod.metadata.labels)
+            else:
+                pod_labels = {}
 
             for container in pod.status.containerStatuses:
                 # add each pod_container to the namespace group, and to each label_value group
@@ -184,8 +185,8 @@ class K8sInventoryHelper(K8sAnsibleMixin):
 
         for service in obj.items:
             service_name = service.metadata.name
-            service_labels = {} if not service.metadata.labels else service.metadata.labels
-            service_annotations = {} if not service.metadata.annotations else service.metadata.annotations
+            service_labels = {} if not service.metadata.labels else dict(service.metadata.labels)
+            service_annotations = {} if not service.metadata.annotations else dict(service.metadata.annotations)
 
             self.inventory.add_host(service_name)
 
@@ -234,7 +235,7 @@ class K8sInventoryHelper(K8sAnsibleMixin):
                 self.inventory.set_variable(service_name, 'load_balancer_ip',
                                             service.spec.loadBalancerIP)
             if service.spec.selector:
-                self.inventory.set_variable(service_name, 'selector', service.spec.selector)
+                self.inventory.set_variable(service_name, 'selector', dict(service.spec.selector))
 
             if hasattr(service.status.loadBalancer, 'ingress') and service.status.loadBalancer.ingress:
                 load_balancer = [{'hostname': ingress.hostname,
@@ -283,8 +284,7 @@ class OpenShiftInventoryHelper(K8sInventoryHelper):
         self.inventory.add_child(namespace_group, namespace_routes_group)
         for route in obj.items:
             route_name = route.metadata.name
-            route_labels = {} if not route.metadata.labels else route.metadata.labels
-            route_annotations = {} if not route.metadata.annotations else route.metadata.annotations
+            route_annotations = {} if not route.metadata.annotations else dict(route.metadata.annotations)
 
             self.inventory.add_host(route_name)
 
@@ -294,6 +294,9 @@ class OpenShiftInventoryHelper(K8sInventoryHelper):
                     group_name = 'label_{0}_{1}'.format(key, value)
                     self.inventory.add_group(group_name)
                     self.inventory.add_child(group_name, route_name)
+                route_labels = dict(route.metadata.labels)
+            else:
+                route_labels = {}
 
             self.inventory.add_child(namespace_routes_group, route_name)
 
@@ -313,4 +316,4 @@ class OpenShiftInventoryHelper(K8sInventoryHelper):
                 self.inventory.set_variable(route_name, 'path', route.spec.path)
 
             if hasattr(route.spec.port, 'targetPort') and route.spec.port.targetPort:
-                self.inventory.set_variable(route_name, 'port', route.spec.port)
+                self.inventory.set_variable(route_name, 'port', dict(route.spec.port))

--- a/lib/ansible/module_utils/k8s/inventory.py
+++ b/lib/ansible/module_utils/k8s/inventory.py
@@ -127,6 +127,9 @@ class K8sInventoryHelper(K8sAnsibleMixin):
             else:
                 pod_labels = {}
 
+            if not pod.status.containerStatuses:
+                continue
+
             for container in pod.status.containerStatuses:
                 # add each pod_container to the namespace group, and to each label_value group
                 container_name = '{0}_{1}'.format(pod.metadata.name, container.name)

